### DISCUSSION TITLE
Improving build system and CI/CD flow

### DISCUSSION THREAD

**Discussion by avinal** [fossology/fossology#1931] [Ideas]
An idea has been proposed to [Migrate the CI from Travis to GitHub Actions](https://github.com/fossology/fossology/wiki/Google-Summer-of-Code-Proposals-2021#migrate-from-travis-to-github-actions) under GSoC 2021. Some tasks are completed. Some more improvements are implementing caching in workflows and created an interconnected workflow system for parallel+serial test/check runs. This will ensure faster and more reliable checks.

Related issues and PR:
- #1913 (Task related to CI listed here)
- #1917 
- #1919 

In the past, there have been [some](https://fossology.slack.com/archives/C6NH0BY5N/p1604490262005300) [discussions](https://fossology.slack.com/archives/C6NH0BY5N/p1604914935002400) on changing the build system of Fossology. There are many possible candidates (`CMake`, `Bazel`, `Ninja`, `Meson` suggestions are welcome). CMake being the most obvious one. However, it needs discussion before deciding. Tagging @mcjaeger @GMishx for more insights into this idea.

Please feel free to make suggestions and improvements :)

**Comment by GMishx** [comment]
This will be a nice addition to FOSSology if we can use new build systems. I do not have experience with any of these build systems but the aim will be
1. The build system should be available on all supported distros.
2. It should be easy to install (like `apt-get install make`)
3. It should definitely improve the build speed.
4. It will be a nice thing if the learning curve for the new build system is not too steep.

Also, for the GitHub Actions task, this issue will be a good add-on: https://github.com/fossology/fossology/issues/207 . What do you think?

**Reply by avinal** [reply]
I did a bit of research and narrowed our choices to 3 (CMake, Meson, and Bazel), however, Bazel uses java runtime and will be an unnecessary overhead if we have no other use in Fossology(Open Source Review Toolkit if integrated will need it), besides it is more Google-like and not very welcomed for other open-source projects. That leaves us to CMake and Meson. 

I have been using CMake for a while now, and TBH it works flawlessly every single time. But it can be very eye-watering to create a CMake configuration for complex projects like this. But once created you can rely on it and is really easy to use. It also has the largest community and availability for distros. (Score: 3/4 very steep learning curve)

Meson is quite new to the market, but looks promising and is [definitely faster](https://thiblahute.github.io/jpakkane.github.io/Simple-comparison.html) or at least equal in terms of speed with CMake. Recently lots of [open source projects](https://mesonbuild.com/Users.html) switched to Meson, it can be installed with `pip` so technically any distro that supports python. Also, the script language of Meson is really easy to learn and understand (score 3.5/4 is too new)

We need a meeting to decide between these two options. 

If I understood correctly, we want to test installation on different distros. Then #207 is definitely a nice addon to the CI task, there is only one problem, GitHub by default provides only Ubuntu, Windows 10, and macOS virtual environment. There can be two solutions to this problem: 
- Self-hosted runners (adds another maintenance task but is highly customizable)
- Dockerised tests with different distros (easy and preferred)

**Reply by GMishx** [reply]
That is a good description there. I agree, a meeting will be better to decide on it.

To answer your queries on #207, you can refer to https://github.com/fossology/fossology/blob/master/.github/workflows/release-publish.yml

**Reply by avinal** [reply]
Ah, I see we are already using dockerized containers. Please add it to the tasks in #1913. This will be a good addition.
